codes/c/chapter_hashing/array_hash_map.c

```diff
@@ -48,162 +69,145 @@ const char *get(const ArrayHashMap *d, const int key) {
-    set->len = total;
+    set->len = index;
 }
 
 /* 获取所有值 */
-void valueSet(ArrayHashMap *d, MapSet *set) {
+void valueSet(ArrayHashMap *d, IndexMembers indexMembers, MapSet *set) {
     char **vals;
     int i = 0, index = 0;
-    int total = 0;
-
-    /* 统计有效键值对数量 */
-    for (i = 0; i < HASH_MAP_DEFAULT_SIZE; i++) {
-        if (d->buckets[i] != NULL) {
-            total++;
-        }
-    }
+    int total = indexMembers.len;
 
     vals = malloc(total * sizeof(char *));
-    for (i = 0; i < HASH_MAP_DEFAULT_SIZE; i++) {
-        if (d->buckets[i] != NULL) {
-            vals[index] = d->buckets[i]->val;
-            index++;
-        }
+    for (i = 0; i <= total; i++) {
+        vals[index] = d->buckets[indexMembers.key[i]]->val;
+        index++;
     }
 
     set->set = vals;
-    set->len = total;
+    set->len = index;
 }
 
 /* 打印哈希表 */
-void print(ArrayHashMap *d) {
+void print(ArrayHashMap *d, IndexMembers indexMembers) {
     int i;
-    MapSet set;
-    pairSet(d, &set);
-    Pair *entries = (Pair *)set.set;
-    for (i = 0; i < set.len; i++) {
+    MapSet *set;
+    set = malloc(indexMembers.len);
+    pairSet(d, indexMembers, set);
+
+    Pair *entries = (Pair *)set->set;
+
+    for (i = 0; i <= set->len; i++) {
         printf("%d -> %s\n", entries[i].key, entries[i].val);
     }
-    free(set.set);
+
+    free(set);
 }
 
 /* Driver Code */
 int main() {
     /* 初始化哈希表 */
     ArrayHashMap *map = newArrayHashMap();
+    IndexMembers index;
+    InitIndex(&index);
 
     /* 添加操作 */
     // 在哈希表中添加键值对 (key, value)
-    put(map, 12836, "小哈");
-    put(map, 15937, "小啰");
-    put(map, 16750, "小算");
-    put(map, 13276, "小法");
-    put(map, 10583, "小鸭");
+    put(map,&index, 12836, "小哈");
+    put(map,&index, 15937, "小啰");
+    put(map,&index, 16750, "小算");
+    put(map,&index, 13276, "小法");
+    put(map,&index, 10583, "小鸭");
     printf("\n添加完成后，哈希表为\nKey -> Value\n");
-    print(map);
+    print(map, index);
 
     /* 查询操作 */
     // 向哈希表输入键 key ，得到值 value
     const char *name = get(map, 15937);
     printf("\n输入学号 15937 ，查询到姓名 %s\n", name);
 
+    /*输出一遍key的位置*/
+    for (int i = 0; i <= index.len; i++) {
+        printf("%d ", hashFunc(index.key[i]));
+    }
+
     /* 删除操作 */
     // 在哈希表中删除键值对 (key, value)
-    removeItem(map, 10583);
+    removeItem(map,&index, 10583);
     printf("\n删除 10583 后，哈希表为\nKey -> Value\n");
-    print(map);
+    print(map, index);
 
     /* 遍历哈希表 */
     int i;
-
-    printf("\n遍历键值对 Key->Value\n");
-    print(map);
-
     MapSet set;
 
-    keySet(map, &set);
+    keySet(map,index, &set);
     int *keys = (int *)set.set;
     printf("\n单独遍历键 Key\n");
     for (i = 0; i < set.len; i++) {
         printf("%d\n", keys[i]);
     }
-    free(set.set);
 
-    valueSet(map, &set);
+    valueSet(map,index, &set);
     char **vals = (char **)set.set;
     printf("\n单独遍历键 Value\n");
     for (i = 0; i < set.len; i++) {
         printf("%s\n", vals[i]);
     }
-    free(set.set);
 
     return 0;
 }
+
```

Comment:
Lack of a free function, memory space is created by `newArrayHashMap` but not freed.

Reply:
Ok, I added now.